src/Npgsql/NpgsqlCommand.cs

```diff
@@ -1259,37 +1236,46 @@ internal async ValueTask<NpgsqlDataReader> ExecuteReader(CommandBehavior behavio
                         // to prevents a dependency on the thread pool (which would also trigger deadlocks).
                         // The WriteBuffer notifies this command when the first buffer flush occurs, so that the
                         // send functions can switch to the special async mode when needed.
-                        var sendTask = NonMultiplexingWriteWrapper(connector, async, finalCt);
+                        var sendTask = NonMultiplexingWriteWrapper(connector, async, writeCt);
 
-                        // The following is a hack. It raises an exception if one was thrown in the first phases
-                        // of the send (i.e. in parts of the send that executed synchronously). Exceptions may
-                        // still happen later and aren't properly handled. See #1323.
-                        if (sendTask.IsFaulted)
-                            sendTask.GetAwaiter().GetResult();
+                        try
+                        {
+                            // The following is a hack. It raises an exception if one was thrown in the first phases
+                            // of the send (i.e. in parts of the send that executed synchronously). Exceptions may
+                            // still happen later and aren't properly handled. See #1323.
+                            if (sendTask.IsFaulted)
+                                sendTask.GetAwaiter().GetResult();
+                        }
+                        catch (OperationCanceledException e)
+                        {
+                            // User requested the cancellation - breaking the connection, as WriteBuffer doesn't do it
+                            throw connector.Break(e);
+                        }
 
                         // TODO: DRY the following with multiplexing, but be careful with the cancellation registration...
                         var reader = connector.DataReader;
                         reader.Init(this, behavior, _statements, sendTask);
                         connector.CurrentReader = reader;
                         if (async)
-                            await reader.NextResultAsync(finalCt);
+                            await reader.NextResultAsync(readCt);
                         else
                             reader.NextResult();
                         return reader;
                     }
-                    catch
+                    catch (Exception e)
                     {
                         connector.CurrentReader = null;
                         conn.Connector?.EndUserAction();
+
+                        if (connector.UserCancellationRequested && e is PostgresException pg && pg.SqlState == PostgresErrorCodes.QueryCanceled)
```

Comment:
Is there any scenario where we get a QueryCanceled PostgresException, but UserCancellationRequested isn't true? If not, we can remove this check (and the property + management around it).

Also, if we always raise QueryCanceled as OperationCanceledException, shouldn't that happen in NpgsqlConnector.ReadMessageLong rather than here?

Reply:
> Is there any scenario where we get a QueryCanceled PostgresException, but UserCancellationRequested isn't true? If not, we can remove this check (and the property + management around it).

Good catch! Yeah, you're right.

> Also, if we always raise QueryCanceled as OperationCanceledException, shouldn't that happen in NpgsqlConnector.ReadMessageLong rather than here?

I was worried, that there might be some checks for the specific exception type, like here:
https://github.com/npgsql/npgsql/blob/a570da94db0e7be3625a024bfbaf3137ad87ad8f/src/Npgsql/NpgsqlDataReader.cs#L260-L270
But it seems to be the only one, so I suppose it should be okay to move the exception handling to the connector.